x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/LifecyclePolicy.java

```diff
@@ -0,0 +1,149 @@
+
+    @SuppressWarnings("unchecked")
+    public static ConstructingObjectParser<LifecyclePolicy, String> PARSER = new ConstructingObjectParser<>("lifecycle_policy", false,
+        (a, name) -> {
+            List<Phase> phases = (List<Phase>) a[0];
+            Map<String, Phase> phaseMap = phases.stream().collect(Collectors.toMap(Phase::getName, Function.identity()));
+            return new LifecyclePolicy(name, phaseMap);
+        });
+    static {
+        PARSER.declareNamedObjects(ConstructingObjectParser.constructorArg(), (p, c, n) -> Phase.parse(p, n), v -> {
+            throw new IllegalArgumentException("ordered " + PHASES_FIELD.getPreferredName() + " are not supported");
+        }, PHASES_FIELD);
+    }
+
+    private final String name;
+    private final Map<String, Phase> phases;
+
+    /**
+     * @param name
+     *            the name of this {@link LifecyclePolicy}
+     * @param phases
+     *            a {@link Map} of {@link Phase}s which make up this
+     *            {@link LifecyclePolicy}.
+     */
+    public LifecyclePolicy(String name, Map<String, Phase> phases) {
+        this.name = name;
+        this.phases = phases;
+        validate(phases.values());
+    }
+
+    public static LifecyclePolicy parse(XContentParser parser, String name) {
+        return PARSER.apply(parser, name);
+    }
+
+    /**
+     * @return the name of this {@link LifecyclePolicy}
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * @return the {@link Phase}s for this {@link LifecyclePolicy} in the order
+     *         in which they will be executed.
+     */
+    public Map<String, Phase> getPhases() {
+        return phases;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.startObject(PHASES_FIELD.getPreferredName());
+        for (Phase phase : phases.values()) {
+            builder.field(phase.getName(), phase);
+        }
+        builder.endObject();
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, phases);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        if (obj.getClass() != getClass()) {
+            return false;
+        }
+        LifecyclePolicy other = (LifecyclePolicy) obj;
+        return Objects.equals(name, other.name) &&
+            Objects.equals(phases, other.phases);
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this, true, true);
+    }
+
+    static void validate(Collection<Phase> phases) {
+        Set<String> allowedPhases = Sets.newHashSet("hot", "warm", "cold", "delete");
+        Map<String, Set<String>> allowedActions = new HashMap<>(allowedPhases.size());
```

Comment:
I don't think we need to rebuild this map every time we validate, can this be a static var and initialized in a `static` block?

Reply:
good call